ServantHDBC/src/Api/Product.hs

```diff
@@ -12,80 +12,113 @@ module Api.Product (ProductAPI, productHandlers) where
 
 import Control.Monad.Trans.Except
 import Data.Aeson
+import Data.Text
+import Data.Time.Clock
 import GHC.Generics
 import Servant
+import Web.HttpApiData
+
 
 
 data Product = Product
-    { ids               :: String
-    , q                 :: String
-    , title             :: String
-    , sku               :: String
-    , prod_type         :: String
+    { id                :: String
+    , tenant_id         :: String
+    , name              :: String
+    , description       :: String
+    , url_slug          :: String
     , tags              :: String
-    , created_at_min    :: String
-    , created_at_max    :: String
-    , updated_at_min    :: String
-    , updated_at_max    :: String
-    , limit             :: String
-    , limit_count       :: Int
-    , offset            :: Int
-    , orderby           :: String
-    , fields            :: String
+    , currency          :: String
+    , advertised_price  :: Int
+    , comparison_price  :: Int
+    , cost_price        :: Int
+    , product_type      :: String
+    , is_published      :: Bool
+    , properties        :: String
     } deriving (Generic)
 
 instance FromJSON Product 
 
-type ProductResult = String
-type ProductAPI = "products" :>
-                (Capture "product_id" Int :>
-                    QueryParam "fields" String :>
+data Field = Field String
+
+instance FromHttpApiData Field where
+    parseUrlPiece x = fmap Field (parseUrlPiece x :: Either Text String)
+
+
+data ProductId = ProductId Int 
+
+instance FromHttpApiData ProductId where
+    parseUrlPiece x = fmap ProductId (parseUrlPiece x :: Either Text Int)
+
+data Sku = Sku String
+
+instance FromHttpApiData Sku where
+    parseUrlPiece x = fmap Sku (parseUrlPiece x :: Either Text String)
+
+data ProductType = ProductType String
```

Comment:
Well, i'd argue that it is. The data declaration alone is enough to demonstrate the types are parsing out correctly. the handler simply takes a product, but it's easy to imagine taking a `(ProductType string)`. I don't know where there is an enumeration of types, but it seems clear the data declaration could be `data ProductType = ProdA | ProdB | ProdC` and the compiler will tell me the handler needs to be updated to handle the 3 variations.


*edit*

Ah, i missed this "If `product_type=physical` then `variants.weight_in_grams` and `variants.weight_display_unit` are required. On the other hand, if `product_type=digital` then the weight related field should NOT be present. Sending weight for a digital product should be an error. And NOT sending weight for a physical product should also be an error."

I'd been avoiding including any logic, just returning strings - i think you're saying you want `data ProductType = Physical | Digital`